lib/authenticated-rest-api/authenticated-rest-api.ts

```diff
@@ -0,0 +1,272 @@
+import * as cdk from "aws-cdk-lib";
+import * as acm from "aws-cdk-lib/aws-certificatemanager";
+
+import { aws_apigateway as apigateway } from "aws-cdk-lib";
+
+import { aws_cloudwatch as cloudwatch } from "aws-cdk-lib";
+import { aws_cloudwatch_actions as cloudwatchActions } from "aws-cdk-lib";
+import { aws_logs as awslogs } from "aws-cdk-lib";
+import { aws_lambda as lambda } from "aws-cdk-lib";
+import { aws_lambda_nodejs as lambdaNodeJs } from "aws-cdk-lib";
+import { Construct } from "constructs";
+import * as path from "path";
+
+import { AuthenticatedRestApiProps } from "./authenticated-rest-api-props";
+import { ResourceProps } from "./resource-props";
+import { buildLambdaEnvironment } from "../util/build-lambda-environment";
+
+const DEFAULT_API_LATENCY_THRESHOLD = cdk.Duration.minutes(1);
+const DEFAULT_LAMBDA_DURATION_THRESHOLD = cdk.Duration.minutes(1);
+
+export class AuthenticatedRestApi extends Construct {
+  readonly restApiId: string;
+  readonly domainName: apigateway.DomainName;
+
+  private restApi: apigateway.RestApi;
+  private alarmAction: cloudwatch.IAlarmAction;
+
+  constructor(scope: Construct, id: string, props: AuthenticatedRestApiProps) {
+    super(scope, id);
+
+    if (
+      (props.domainName && !props.certificateArn) ||
+      (!props.domainName && props.certificateArn)
+    ) {
+      cdk.Annotations.of(scope).addError(
+        `To use a custom domain name both certificateArn and domainName must be specified`,
+      );
+    }
```

Comment:
The conditional looks quite complicated when compared to the statement in the message. Are they both mandatory? Or is it both or neither?

Reply:
Neither are mandatory, but if one is specified, they both have to be.

Reply:
Maybe not. They are both mandatory, so this can be removed.